### DISCUSSION TITLE
Continue model tuning starting from 'model-best'

### DISCUSSION THREAD

**Discussion by Joanna-Wojdylo** [explosion/spaCy#9824] [Help: Coding & Implementations] [answered]
Hi Team, 

I have a question, which is closely related with [https://github.com/explosion/spaCy/discussions/9725](https://github.com/explosion/spaCy/discussions/9725): I want to train my NER model iteratively, starting from en_core_web_trf then adding new data, treat obtained 'model-best' as a new start, train model-best with new data, save it as next checkpoint and so on. As advised I changed [components.ner] `factory = "ner"` to `source = "<path to best-model>"`. My question is: is it all that needs to be done in order to use model-best not en_core_web_trf in future training? Is there any "smart" way to check what exactly is used under CLI train command? I want to make sure, that I am not starting from clear en_core_web_trf every time I train model on new data portion.  

I'm pasting my config file, could you please check if it's correct to start training with model-best?

```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "en"
pipeline = ["transformer","ner"]
batch_size = 128
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
source = "/content/model-best"

incorrect_spans_key = null
moves = null
scorer = {"@scorers":"spacy.ner_scorer.v1"}
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = false
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "*"

[components.transformer]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v3"
name = "roberta-base"
mixed_precision = false

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96

[components.transformer.model.grad_scaler_config]

[components.transformer.model.tokenizer_config]
use_fast = true

[components.transformer.model.transformer_config]

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0
ents_per_type = null

[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

Any tips and explanation will be appreciated :) Thank you!

**Comment by adrianeboyd** [comment, marked as answer]
Hi, you need to source both the `transformer` and `ner` components to continue updating from a trained model. The config above resets the transformer weights for `roberta-base`, which isn't what you want.

**Reply by Joanna-Wojdylo** [reply]
Do I understand you correctly, should I do the same with factory for [components.transformer] and that's it? Or do I need to change anything more, eg. `name="roberta-base"` to something else? Is there any way to track what model is currently trained?

Also, is it OK that I need to install spacy-transformers library or is it a sign that my config file is incorrect?

**Reply by adrianeboyd** [reply]
You leave out all the other settings when you use `source` because they're read in from the sourced pipeline:

```ini
[components.tagger]
source = "/path/to/model-best"

[components.transformer]
source = "/path/to/model-best"
```

When this is saved (after `train` or `assemble`), you'll see the original `roberta-base` references in the saved config (although the name isn't actually used again after initialization because all the parameters/weights are saved directly in `model-best`).

You have to install `spacy-transformers` to use any `transformer` component.

**Reply by Joanna-Wojdylo** [reply]
Thank you! I'll mark this discussion as answered
